clients/include/testing_syr2.hpp

```diff
@@ -0,0 +1,216 @@
+        CHECK_HIP_ERROR(dA_1.memcheck());
+        CHECK_HIP_ERROR(dx.memcheck());
+        CHECK_HIP_ERROR(dy.memcheck());
+
+        EXPECT_ROCBLAS_STATUS(
+            rocblas_syr2<T>(handle, uplo, N, &h_alpha, dx, incx, dy, incy, dA_1, lda),
+            (N < 0 || lda < N || lda < 1 || !incx || !incy) ? rocblas_status_invalid_size
+                                                            : rocblas_status_success);
+
+        return;
+    }
+
+    size_t abs_incx = incx >= 0 ? incx : -incx;
+    size_t abs_incy = incy >= 0 ? incy : -incy;
+    size_t size_A   = size_t(lda) * N;
+    size_t size_x   = N * abs_incx;
+    size_t size_y   = N * abs_incy;
+
+    // Naming: dK is in GPU (device) memory. hK is in CPU (host) memory
+    host_vector<T> hA_1(size_A);
+    host_vector<T> hA_2(size_A);
+    host_vector<T> hA_gold(size_A);
+    host_vector<T> hx(size_x);
+    host_vector<T> hy(size_y);
+    host_vector<T> halpha(1);
+    halpha[0] = h_alpha;
+
+    // allocate memory on device
+    device_vector<T> dA_1(size_A);
+    device_vector<T> dA_2(size_A);
+    device_vector<T> dx(size_x);
+    device_vector<T> dy(size_y);
+    device_vector<T> d_alpha(1);
+    CHECK_HIP_ERROR(dA_1.memcheck());
+    CHECK_HIP_ERROR(dA_2.memcheck());
+    CHECK_HIP_ERROR(dx.memcheck());
+    CHECK_HIP_ERROR(dy.memcheck());
+
+    double gpu_time_used, cpu_time_used;
+    double rocblas_gflops, cblas_gflops, rocblas_bandwidth;
+    double rocblas_error_1;
+    double rocblas_error_2;
+
+    // Initial Data on CPU
+    rocblas_init(hA_1, true);
+    rocblas_init(hx, false);
+    rocblas_init(hy, false);
+
+    hA_2    = hA_1;
+    hA_gold = hA_1;
+
+    CHECK_HIP_ERROR(dA_1.transfer_from(hA_1));
+    CHECK_HIP_ERROR(dx.transfer_from(hx));
+    CHECK_HIP_ERROR(dy.transfer_from(hy));
+
+    if(arg.unit_check || arg.norm_check)
+    {
+        // copy data from CPU to device
+        CHECK_HIP_ERROR(dA_2.transfer_from(hA_2));
+        CHECK_HIP_ERROR(d_alpha.transfer_from(halpha));
+
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+        CHECK_ROCBLAS_ERROR(
+            rocblas_syr2<T>(handle, uplo, N, &h_alpha, dx, incx, dy, incy, dA_1, lda));
+
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device));
+        CHECK_ROCBLAS_ERROR(
+            rocblas_syr2<T>(handle, uplo, N, d_alpha, dx, incx, dy, incy, dA_2, lda));
+
+        // copy output from device to CPU
+        CHECK_HIP_ERROR(hA_1.transfer_from(dA_1));
+        CHECK_HIP_ERROR(hA_2.transfer_from(dA_2));
+
+        // CPU BLAS
+        cpu_time_used = get_time_us();
+        cblas_syr2<T>(uplo, N, h_alpha, hx, incx, hy, incy, hA_gold, lda);
+        cpu_time_used = get_time_us() - cpu_time_used;
+        cblas_gflops  = syr2_gflop_count<T>(N) / cpu_time_used * 1e6;
+
+        if(arg.unit_check)
+        {
+            unit_check_general<T>(N, N, lda, hA_gold, hA_1);
+            unit_check_general<T>(N, N, lda, hA_gold, hA_2);
```

Comment:
Use `unit_check_general<T, T>()` to resolve compilation errors. An extra template argument was added, to allow `float` and `rocblas_bfloat16` to be compared, allowing the `rocblas_bfloat16` to either equal the rounded or truncated value of the `float`. But this is only used in `gemm_ex`, and most functions should simply use two equal type arguments.

Reply:
Lee I presume you didn't set a default template value U=T so something isn't overlooked?

Reply:
Default template arguments won't work. The reason for the explicit template arguments is to force the `host_vector` / `device_vector` arguments into pointers without having to use explicit casts or function calls.

By explicitly specifying the template arguments, which are used as `T*` and `U*` in the function parameter declarations, any `device_vector` or `host_vector` passed as an argument gets implicitly converted to a pointer to its data, while if even if we used default template arguments, such as
```
template<typename T, typename U = T>
void unit_check_general(T* x, U*x, ...)
{
}
```
Then a call of:

```
unit_check_general<T>(vec1<T>, vec2<U>);
```
would try to deduce `U*` from the  `vec2<U>` argument, and would fail. By specifying:
```
unit_check_general<T, U>(vec1<T>, vec2<U>);
```
we force the `T` and `U` to be explicitly specified and not deduced, and then the `T*` and `U*` parameters will force their `device_vector` or `host_vector` arguments to be converted to the pointers to their data.

Reply:
Note that a _failed deduction_ is not the same as a _non-deduced context_. The above uses `U*` in a _deduced context_, and that deduction fails if we pass an argument of type `device_vector` or `host_vector`.

If we wanted to make the second template argument optional, while keeping the functionality, we could do it this way:

```
template<typename T, typename U = T>
void unit_check_general(T* x, std::type_identity_t<U>* y)
{
}
```
`std::type_identity_t` is new to C++20, but it's trivial class we could implement ourselves:
```
template<typename T>
struct type_identity
{ 
   using type = T;
};
template<typename T>
using type_identity_t = typename type_identity<T>::type;
```
The reason this works is that `std::type_identity_t<U>` is a _non-deduced context_ for `U`, and since there are no _deduced contexts_ for `U` in the function parameter list, the default template argument definition of `U = T` can apply.

The way the code is written now, with `unit_check_general(T* x, U* y)`, both `T` and `U` are in deduced contexts, and hence if we were to specify a default `U = T`, it would be ignored.

I'm just rambling on in language-legaleze to show that I've thought a lot about it.

It's simpler now to just require both arguments to be explicitly specified. If this were to break users of a library who were used to passing a single type argument to `unit_check_general`, then we should go out of our way to implement the non-deduced context with `type_identity_t`, so that if `U` is not explicitly specified, it defaults to `T`.

Reply:
Figured there was a reason, good to know.   Thanks.